### PR TITLE
NN II Revisions

### PR DESCRIPTION
**Pull request by trgardos**
Revised and reformatted for slides. It's 84 slides however, but some are basically build, but it may overflow 1 lecture.

I think NN-III might shorter than 75 minutes, so i think it is ok to have NN-II run into the 2nd lecture.

### CODE REVIEW

File: ds701_book/24-NN-II-Backprop.qmd [hunk 1]
```diff
@@ -145,79 +120,84 @@ plt.grid()
...
+$$
+
+which is calculated for each training example $i$ and then accumulated.
+
+So what we want to do is find the partial derivatives of $L$ with respect to each parameter in the network.
+
+In other words, we want to compute
+
+$$
+\frac{\partial L}{\partial \mathbf{\beta}_k}, \frac{\partial L}{\partial \mathbf{\Omega}_k}, \frac{\partial L}{\partial \mathbf{\beta}_{k-1}}, \frac{\partial L}{\partial \mathbf{\Omega}_{k-1}}, \ldots, \frac{\partial L}{\partial \mathbf{\beta}_0}, \frac{\partial L}{\partial \mathbf{\Omega}_0}
```
Comment (sladenheim) [inline ds701_book/24-NN-II-Backprop.qmd:189]: We haven't mentioned derivatives with respect to vectors or matrices, is it worth remarking here that there are well-defined rules that govern how these are calculated and that, e.g., pytorch, handle these calculations for you?

File: ds701_book/24-NN-II-Backprop.qmd [hunk 2]
```diff
@@ -145,79 +120,84 @@ plt.grid()
...
+E.g. a regression problem.
+
+## FCN Matrix Equations
+
+We can express the matrix equations for a $K$ layer network as
+
+$$
+\begin{aligned}
+\mathbf{h}_1 &= \mathbf{a}[\mathbf{\beta}_0 + \mathbf{\Omega}_0 \mathbf{x}] \\
+\mathbf{h}_2 &= \mathbf{a}[\mathbf{\beta}_1 + \mathbf{\Omega}_1 \mathbf{h}_1] \\
```
Comment (sladenheim) [inline ds701_book/24-NN-II-Backprop.qmd:169]: I think the \beta_i should be boldface since they are vectors.

File: ds701_book/24-NN-II-Backprop.qmd [hunk 3]
```diff
@@ -32,74 +34,41 @@ With that we can implement a neural network training loop, and see how similar i
...
-$$y = \beta_0 + \beta_1 u + \beta_2 v$$
+The **sigmoid function** is
 
-We just renamed things: $\beta_0 \gets b$, $\beta_1 \gets w_0$, $\beta_2 \gets w_1$, $u \gets x_0$ and $v \gets x_1$ for $i \in [0,1]$.
-
-> So multiple linear regression can be viewed as one linear neuron.
-
-Activation function is typically some nonlinear function that compresses the input in some way. Historically, it's been the sigmoid and $\tanh()$ functions. See for example [Hyperbolic Functions](https://en.wikipedia.org/wiki/Hyperbolic_functions#Tanh).
+$$
+\mathrm{sigmoid}(x) = \frac{1}{1 + e^{-x}}
```
Comment (sladenheim) [inline ds701_book/24-NN-II-Backprop.qmd:67]: Period at the end of this formula.

File: ds701_book/24-NN-II-Backprop.qmd [hunk 4]
```diff
@@ -119,23 +88,29 @@ plt.show()
 
 The hyperbolic tangent, $\tanh$, is basically the sigmoid function shifted and scaled to a range of [-1,1].
 
+The **tanh** function is
+
+$$
+\tanh(x) = \frac{e^x - e^{-x}}{e^x + e^{-x}}
```
Comment (sladenheim) [inline ds701_book/24-NN-II-Backprop.qmd:94]: Period at the end of this formula.

File: ds701_book/24-NN-II-Backprop.qmd [hunk 5]
```diff
@@ -145,79 +120,84 @@ plt.grid()
...
 
 Multiple artificial neurons can be acting on the same inputs, in what we call
 a _layer_, and we can have more than one _layer_ until we produce one or more
 outputs.
 
-The example above shows a network with _3 inputs_, two layers of neurons, each
-with 4 neurons, followed by one layer that produces a single value output.
+The example above shows a network with _3 inputs_, three hidden layers of 4, 2 and 3 neurons, respectively, followed by one layer that produces two continuous value outputs.
+
+E.g. a regression problem.
```
Comment (sladenheim) [inline ds701_book/24-NN-II-Backprop.qmd:160]: E.g.,

File: ds701_book/24-NN-II-Backprop.qmd [hunk 6]
```diff
@@ -367,6 +353,7 @@ In order to calculate the gradients, we will need to capture the computation gra
 To start with, we'll accept a tuple of child nodes in the initializer and store that as a set in the object.
 
 ```{python}
+#| code-fold: false
 # Value version 4
 class Value:
                         #    vvvvvvvvvvvv
```
Comment (sladenheim) [inline ds701_book/24-NN-II-Backprop.qmd:359]: I think this is a typo.

File: ds701_book/24-NN-II-Backprop.qmd [hunk 7]
```diff
@@ -755,11 +763,15 @@ How do we get $\partial{L}/\partial{c}$?
 
 To paraphrase from the Wikipedia page on [Chain rule](https://en.wikipedia.org/wiki/Chain_rule), if a variable $L$ depends on the variable $e$, which itself depends on the variable $c$ (that is, $e$ and $L$ are dependent variables), then $L$ depends on $c$ as well, via the intermediate variable $e$. In this case, the chain rule is expressed as
```
Comment (sladenheim) [inline ds701_book/24-NN-II-Backprop.qmd:764]: I think we may want to use a proper calculus book citation, such as James Stewart, Calculus Early Transcendentals. Or we can just assume a general knowledge of the chain rule.

File: ds701_book/24-NN-II-Backprop.qmd
Comment (sladenheim) [inline ds701_book/24-NN-II-Backprop.qmd:1]: A Single Neuron section needs a space to properly render the list. This is on line 960.

File: ds701_book/24-NN-II-Backprop.qmd [hunk 8]
```diff
@@ -967,25 +996,23 @@ differentiate the ReLU.
...
 
 ```{python}
-#| tags: []
+#| code-fold: false
 o.grad = 1.0
 ```
 
 <!-- Credit: _Understanding Deep Learning_, Figure 7.6 -->
 ReLU is technically not differentiable at 0, but practically we implement the
 derivative as 0 when $ \le 0$ and 1 when $ 1 > 0 $
```
Comment (sladenheim) [inline ds701_book/24-NN-II-Backprop.qmd:1010]: Need to remove these spaces for the latex to render properly.

File: ds701_book/24-NN-II-Backprop.qmd [hunk 9]
```diff
@@ -1146,38 +1175,44 @@ The risk now is that if you don't zero the gradients for the next update iterati
...
-\end{aligned}$$
+\end{aligned}
+$$
 
 So the partial derivative $\frac{\partial f}{\partial a}$ for the value $a = -2.0$ is $-3.0$.
-
-```
+:::
 
 ## Enhancements to `Value` Class
```
Comment (sladenheim) [inline ds701_book/24-NN-II-Backprop.qmd:1218]: Need a space after We addded (see below) for the list to render properly.

File: ds701_book/24-NN-II-Backprop.qmd [hunk 10]
```diff
@@ -1425,7 +1460,7 @@ class MLP(Module):
 ```
 
 ```{python}
-#| tags: []
+#| code-fold: false
 # help(Module)
```
Comment (sladenheim) [inline ds701_book/24-NN-II-Backprop.qmd:1464]: Why is all this code commented out?

**Review by sladenheim — CHANGES_REQUESTED** [review]
I know you're still working on this draft but I spotted a few things that could be changed.

One general comment that I didn't want to add everywhere, but we should be using punctuation after formulas.

File: ds701_book/24-NN-II-Backprop.qmd [hunk 11]
```diff
@@ -145,79 +120,84 @@ plt.grid()
...
-### Relating Back to _Another_ Earlier Lecture
+## Relating Back to an Earlier Lecture
 
 __Question__
 
 What does 
 
-$$ \mathrm{sigmoid}(\sum_i w_i x_i + b) \hspace{10pt} \textrm{where} \hspace{10pt} \mathrm{sigmoid}(x) = \frac{1}{1 + e^{-x}}$$
+$$
+\mathrm{sigmoid}(\sum_i w_i x_i + b) \hspace{10pt} \textrm{where} \hspace{10pt} \mathrm{sigmoid}(x) = \frac{1}{1 + e^{-x}}
```
Comment (sladenheim) [inline ds701_book/24-NN-II-Backprop.qmd:130]: Use \left( and \right) around the sum to make the parentheses match the size of the sum.

File: ds701_book/24-NN-II-Backprop.qmd (hunk 7 above)
Comment (trgardos) [reply ds701_book/24-NN-II-Backprop.qmd:764]: have you looked at the wikipedia page? It's actually quite a good reference. Is Stewart available online?

File: ds701_book/24-NN-II-Backprop.qmd (hunk 5 above)
[the remark this answers is not in the document]
Comment (trgardos) [reply ds701_book/24-NN-II-Backprop.qmd:160]: Addedf "This is regression model that is predicting two continuous valued
outputs."

File: ds701_book/24-NN-II-Backprop.qmd (hunk 1 above)
Comment (trgardos) [reply ds701_book/24-NN-II-Backprop.qmd:189]: Added

In the notation above, $\mathbf{\beta}_k$ and $\mathbf{\Omega}_k$ are the bias
vectors and weight matrices of the $k$-th layer. Just as with partial derivatives
with respect to scalars, there are well-defined rules for how to compute the
partial derivatives of vector and matrix expressions and, in fact, PyTorch and
TensorFlow implement these calculations.

File: ds701_book/24-NN-II-Backprop.qmd (hunk 2 above)
Comment (trgardos) [reply ds701_book/24-NN-II-Backprop.qmd:169]: I do have them marked as boldface but \beta doesn't appear to render that way. Ideas?

File: ds701_book/24-NN-II-Backprop.qmd (hunk 6 above)
Comment (trgardos) [reply ds701_book/24-NN-II-Backprop.qmd:359]: which? the vvvv's are pointing to the new argument, but I'm going to consolidate these to fewer versions anyway. stay tuned.

File: ds701_book/24-NN-II-Backprop.qmd (hunk 10 above)
Comment (trgardos) [reply ds701_book/24-NN-II-Backprop.qmd:1464]: I didn't want to create a bunch of output but encouraging students to uncomment and try. Added

"Feel free to uncomment lines of code in the following cell to see the docstrings for each class."

I might still remove this when I consolidate.

**Comment by trgardos** [comment]
> I know you're still working on this draft but I spotted a few things that could be changed.
> 
> One general comment that I didn't want to add everywhere, but we should be using punctuation after formulas.

Thanks. I appreciate the feedback. And yes, I will try to remember to punctuate equations. :-)

**Comment by trgardos** [comment]
Ok. I finished revisions and now just have to update slide formatting. Will do that this afternoon.

File: ds701_book/24-NN-II-Backprop.qmd (hunk 2 above)
Comment (sladenheim) [reply ds701_book/24-NN-II-Backprop.qmd:192]: For symbols like \beta, you need \boldsymbol{\beta}

File: ds701_book/24-NN-II-Backprop.qmd (hunk 7 above)
Comment (sladenheim) [reply ds701_book/24-NN-II-Backprop.qmd]: I don't think Stewart is (freely) available online. I think this is fine. Can always revisit at another time since we use it in previous lectures.

File: ds701_book/24-NN-II-Backprop.qmd (hunk 10 above)
Comment (sladenheim) [reply ds701_book/24-NN-II-Backprop.qmd:1662]: No problem. I think this is fine.

File: ds701_book/24-NN-II-Backprop.qmd (hunk 2 above)
Comment (trgardos) [reply ds701_book/24-NN-II-Backprop.qmd:192]: oh! thanks!

**Review by sladenheim — APPROVED** [review]
I committed changes to this branch that edited language and fixed typos.

**Comment by trgardos** [comment]
> I committed changes to this branch that edited language and fixed typos.

Thank you!
